### ISSUE TITLE
Low threshold setting very high when near agile peak rates

### ISSUE THREAD

**Issue by Philhews** [springfall2008/batpred#367]
Once I've got within 4 hours of the agile peak rates, the low threshold has jumped up to equal those high rates
![image](https://github.com/springfall2008/batpred/assets/117396644/4603cdaa-5a8b-4be8-864a-0e9757fa350e)
Now I have almost all of my periods defined as possible charge periods, I don't think it will cause any problems but seems fairly odd.
![image](https://github.com/springfall2008/batpred/assets/117396644/81aab06f-0a38-47b9-bd76-06adbab3aa8c)
am on v7.13.8 if that matters

**Comment by gcoan** [comment]
I'm seeing the same behaviour, also 7.13.8

![image](https://github.com/springfall2008/batpred/assets/142018870/085671c5-7187-4c05-89e9-7502195b13ad)

I'm wonder if this is why predbat is favouring discharging at 28p over using it to supply home demand and avoiding grid import at ~42p?

The plan at 14:31
![image](https://github.com/springfall2008/batpred/assets/142018870/4cc61b20-4f84-4162-a933-cf94c7613492)

And then plan at 15:04
![image](https://github.com/springfall2008/batpred/assets/142018870/385246e2-fae9-4bb4-a921-820895a6a34b)

And at 15:14
![image](https://github.com/springfall2008/batpred/assets/142018870/f254e6d3-1cc1-4fd7-98cc-1f069d6507cf)

Quite a bit of change in the plan but in all versions there's import in peak rate and peak rate discharge as well when it would be cheaper to just supply the house from the battery

**Comment by Philhews** [comment]
It seems it's something to do with the future rates too because when I change my future rate metric down from 6 to 0 it gets better maybe? Certainly lower but only gives a handful of potential charge slots
![Screenshot_2023-11-24-17-20-46-82_c3a231c25ed346e59462e84656a70e50](https://github.com/springfall2008/batpred/assets/117396644/70a80d86-8022-48be-829c-70a59b49596f)

**Comment by slopemad** [comment]
I think a recent change in Predbat involves taking into account all the slots in the next 4 hours which might cause this effect.  What I did notice is that Predbat instructed my inverter to go into 'freeze charging' at 16:20 (and back to Idle at 16:30), and thus stopped running off battery and instead off mains during super peak rate for 10 minutes :-(

I also note my Low Threshold shot up at about 12:35, momentarily dropping much lower at about 13:40, 16:00 and 1630(ish), as can be seen in the below chart.

<img width="601" alt="Screenshot 2023-11-24 at 17 36 25" src="https://github.com/springfall2008/batpred/assets/118896510/774c1b6c-b0ae-4fd5-ae0d-09409bf5af3e">

<img width="920" alt="Screenshot 2023-11-24 at 17 38 59" src="https://github.com/springfall2008/batpred/assets/118896510/48927d58-89b9-40af-8acb-4a2cb8c475f9">

**Comment by gcoan** [comment]
I don't have future rate metric set to anything, just relying on the nord data and rollover of current rates.

Because the plan remained to keep charging in the peak period I had to turn predbat into read only mode and manually change the inverter settings to start supporting the house from the battery instead of the grid.
It's still in read only mode but it looks like this low rate is causing predbat to continue to want to charge the battery despite the rates continuing to fall through the evening. The better plan would be to allow running off grid import until 22:30-23:30, and charge in this period to give sufficient soc to get over the following more expensive hour of import

![image](https://github.com/springfall2008/batpred/assets/142018870/b9153a6c-9348-41a6-a719-43c819be6cde)

**Comment by gcoan** [comment]
My plan has a similar pattern of the low rate threshold jumping up as we approach the peak period but having a couple of jumps up and down beforehand

**Comment by slopemad** [comment]
What is your best_soc_keep set to?

**Comment by slopemad** [comment]
OK, so looking at your plan, while it says 'Charge', it isn't actually charging, the SOC is higher than the limit and so is continuing to discharge the battery down to 14% (is that where your best_soc_keep or even best_soc_min is set to?).
At 19:30, it then wants to charge up at 24p so it can discharge when the rate is 27p.  Similar at 20:30.  I seem to recall you changed your settings to not take into account the cost of using up battery cycles, hence charge at 27 to discharge at 24.

I have the battery cycle cost set to 3p, there's not much value (to me) in charging at 24p to discharge when rates are 27p.

I think the low threshold will drop after 7pm when the peak is no longer in that 4 hour window, I guess that particular concept needs some refinement.

<img width="441" alt="Screenshot 2023-11-24 at 18 45 27" src="https://github.com/springfall2008/batpred/assets/118896510/62d1579f-03cf-47fc-b4c2-4da5c034c33c">

**Comment by springfall2008** [comment]
I'd hazard a guess that metric_keep is set to 14% or min_soc is set to this level and hence why it's holding at the battery at that level?

**Comment by slopemad** [comment]
This plan just doesn't seem right?  It's now planning to do it's bulk charge now, at 22p, rather than potentially just holding the charge level now and charging at 18p or 19p later in the night.  And maintaining that charge until the morning and exporting excess solar (high solar forecast for tomorrow).  It just feels like the logic has messed up somehow?

<img width="573" alt="Screenshot 2023-11-24 at 20 45 26" src="https://github.com/springfall2008/batpred/assets/118896510/0091c622-7f46-48ce-82d1-dde4a3f2d16e">

<img width="556" alt="Screenshot 2023-11-24 at 20 51 25" src="https://github.com/springfall2008/batpred/assets/118896510/1a8622cb-d1b1-4ab5-9387-5f514bdf1672">

**Comment by gcoan** [comment]
> OK, so looking at your plan, while it says 'Charge', it isn't actually charging, the SOC is higher than the limit and so is continuing to discharge the battery down to 14% (is that where your best_soc_keep or even best_soc_min is set to?).
> 
Yes agree, some of the labelling of "charge" slots can take some interpretation. I've seen charge slots actually being "hold charging"

My best soc keep is set to 0.3kW which on a 5.2 battery is 5.7% - only just above the 4% reserve figure. Best soc min and max is 0, so those shouldn't be influencing the battery retention.

> At 19:30, it then wants to charge up at 24p so it can discharge when the rate is 27p.  Similar at 20:30.  I seem to recall you changed your settings to not take into account the cost of using up battery cycles, hence charge at 27 to discharge at 24.
> 
> I have the battery cycle cost set to 3p, there's not much value (to me) in charging at 24p to discharge when rates are 27p.

Agree. I've tended to manage that with the inverter charge and discharge losses (currently 4 & 4%) so it's marginal benefit at best to charge at 19:30 at 24p, bringing the soc up from 14% if the predbat plan is to then discharge that at 27p from 20:00.  However the plan at 20:0 is also labelled as "charge" which didn't make sense, but looking at the soc it's shown as dropping from 38% so maybe it's mislabelled and it's discharging not charging ??

Maybe the 20:30 charge is calculated as cost effective for the same reasons; 24.26p gets rounded to 24p and the 21:00 slot is 25.74p rounded to 26p which is just on the limit of my inverter loss difference so it works out to be cost effective even though the unrounded figures wouldn't be!

> I think the low threshold will drop after 7pm when the peak is no longer in that 4 hour window, I guess that particular concept needs some refinement.

Low rate threshold did indeed drop but it's still above all the immediate next slots so I'm guess why it keeps charging to 100% even though the subsequent slots are dropping in rate and grid import would be cheaper 

![image](https://github.com/springfall2008/batpred/assets/142018870/b37771d1-5690-42cb-b3b0-79c1bf17ab1b)
![image](https://github.com/springfall2008/batpred/assets/142018870/8db30639-bf1f-429d-a3d6-e0e1c91e054c)

**Comment by gcoan** [comment]
My best soc keep is 0.3kW which is 5.7% of the 5.2 battery, only just above 4% reserve

Best soc min is 0

**Comment by slopemad** [comment]
So, there was the plan I posted 23 minutes ago, and it's now changed to something more sensible.  But I've noticed the low and high thresholds are up and down like an... like an uppy downy thing.  I can't see anything in the appdaemon logs which may suggest it is struggling to get rates from the Octopus Energy entities and everything looks consistent in there...

<img width="477" alt="Screenshot 2023-11-24 at 21 22 09" src="https://github.com/springfall2008/batpred/assets/118896510/3838f588-eb28-458f-a79c-5e584b6a81e1">

<img width="449" alt="Screenshot 2023-11-24 at 21 16 10" src="https://github.com/springfall2008/batpred/assets/118896510/53c813e6-4831-44dd-a381-9dfd2e42b3ec">

**Comment by gcoan** [comment]
There must be some sort of "hang-over" in setting the low rate threshold after the peak period.

On 7.13.8:
![image](https://github.com/springfall2008/batpred/assets/142018870/89daa64d-9428-40cd-bf48-6123e5844fbd)

Upgraded to 7.13.9, nothing else changed:
![image](https://github.com/springfall2008/batpred/assets/142018870/32d881aa-91cb-4902-ba97-3a0cfedd38fc)

Sometimes the import rates are shown green, sometimes pale yellow and I've seen the same rate change colour when the plan is recalculated?

I then on 7.13.9 increased the charge & discharge losses from 4 to 5% and the new plan looks better so I'll be turning predbat back off read only after 10pm
![image](https://github.com/springfall2008/batpred/assets/142018870/83ff8595-81e5-41d5-9675-41a129fd1add)

Some of the charging labels look to be incorrect if I look at the soc direction and soc limit compared to previous slot. Eg 01:00am

**Comment by slopemad** [comment]
Continuing to see fluctuating low threshold and high threshold rates (meaning that Predbat keeps choosing expensive charging slots).  
<img width="318" alt="Screenshot 2023-11-25 at 09 02 54" src="https://github.com/springfall2008/batpred/assets/118896510/aba13fcd-2434-4c92-a6e0-4b11fec41b72">
<img width="321" alt="Screenshot 2023-11-25 at 09 03 03" src="https://github.com/springfall2008/batpred/assets/118896510/56cf034d-b842-4ee9-9629-fb0250199fff">
<img width="1059" alt="Screenshot 2023-11-25 at 09 04 47" src="https://github.com/springfall2008/batpred/assets/118896510/023aac46-785c-4aaa-8656-f1133dc10d47">

**Comment by gcoan** [comment]
There was fluctuations in my high rate threshold overnight but the low rate has been pretty constant, just going up a touch at about 6am.

![image](https://github.com/springfall2008/batpred/assets/142018870/c20ccc20-1881-42ed-86d7-7a399a9c9b47)

If predbat sticks to this plan and keeps the battery charged to 100% up to 4pm peak then I'll be happy with this 
![image](https://github.com/springfall2008/batpred/assets/142018870/0273241d-8fd1-4174-a6dc-9d6832c2441f)

7.13.9

**Comment by gcoan** [comment]
The "looking ahead 4 hours" recently added is causing side effects.

The plan earlier which looked ok, retaining 100% soc to the start of the peak period

At 12:15 the plan was showing charging in the first 4pm peak rate slot
![image](https://github.com/springfall2008/batpred/assets/142018870/80aede4e-0c13-4e2d-b28d-702d598a19f2)

And at 13:25 the charging now extends to 17:30 with unnecessary peak rate discharges.  I added the low rate threshold to the dashboard and can see that that's what's throwing the plan off
![image](https://github.com/springfall2008/batpred/assets/142018870/0561a708-f771-4825-a8d5-d327371f7621)

This may also be the cause of #309 and #357

I'll have to turn predbat off and manually back to eco mode at 4pm

**Comment by gcoan** [comment]
Hmm.  Maybe not as simple as the 4 hours lookahead causing the peak period charging. 

4:10 plan on 7.13.9
![image](https://github.com/springfall2008/batpred/assets/142018870/5eb90d39-3b69-4602-b418-9f91829a2228)

Updated to 7.13.10 and the low rate threshold has dropped but still retaining 100% soc through the peak 🤷‍♂️
![image](https://github.com/springfall2008/batpred/assets/142018870/6f34280a-6e6e-45ca-ab23-f97fa82dc60e)
![image](https://github.com/springfall2008/batpred/assets/142018870/989b6a53-45a3-4728-bdb7-7dab8ab279cd)

Looks like the low rate threshold dropping was temporary because it's just gone back up again 
![image](https://github.com/springfall2008/batpred/assets/142018870/effd43db-d2d5-47f9-acef-16224995378c)
![image](https://github.com/springfall2008/batpred/assets/142018870/ed0b31df-cf71-4f8e-9514-7468314ee2e9)

**Comment by gcoan** [comment]
Excited to see @springfall2008 had released 7.13.11 with "changes to metric keep to avoid near term force charges"

But half good, half not good.

At the moment I've left predbat in read only through the peak and was going to leave it off until the rates drop later this evening.

Plan on 7.13.10 :
![image](https://github.com/springfall2008/batpred/assets/142018870/b5820d1f-d137-473f-9c15-1bc1ea512f3a)
![image](https://github.com/springfall2008/batpred/assets/142018870/50be679a-e217-43e1-bacb-070e46a107d6)

Installed 7.13.11 and good news the peak rate charging and charging this evening when it's not beneficial has disappeared 
![image](https://github.com/springfall2008/batpred/assets/142018870/efde89f7-d9ca-4a1d-9f5a-60c94d41d542)

Bad news is predbat doesn't try to do any charging at all (other than a couple of slots), leaves the battery empty and running off grid tomorrow and into tomorrow's peak.
I know the rates are not great overnight but charging from say 3-5am would be beneficial, and certainly need to charge before tomorrow evenings peak
![image](https://github.com/springfall2008/batpred/assets/142018870/d4dabdeb-7b35-40bd-b71c-f6d20b8285af)
![image](https://github.com/springfall2008/batpred/assets/142018870/4411b5f5-0f25-4981-9c6d-fc963066f93a)

**Comment by springfall2008** [comment]
Did you try setting best soc keep to 0.3 as per the new proposal?

**Comment by gcoan** [comment]
> Did you try setting best soc keep to 0.3 as per the new proposal?

I already had it set to 0.3.  Sorry !

Here's the logfile if it gives any help
[appdaemon (17).log](https://github.com/springfall2008/batpred/files/13464984/appdaemon.17.log)

**Comment by springfall2008** [comment]
Can you try pushing keep up a bit and see what changes? 

From what I can tell it’s on a cusp of being not worth using the battery

**Comment by gcoan** [comment]
> Can you try pushing keep up a bit and see what changes?
> 
> From what I can tell it’s on a cusp of being not worth using the battery

Will do.  I could see that the rates were pretty flat and with 5% charge loss % 5% discharge loss using the battery would be heading towards marginal, but still thought some overnight charging would be worthwhile

**Comment by Philhews** [comment]
What's the reason to recommend 0.3 for the keep metric?

**Comment by springfall2008** [comment]
It’s just above the reserve so you can use the whole battery but steering away from not using it

**Comment by gcoan** [comment]
best_soc_keep = 0.4
the import rates all turned to green from yellow and red (assume this means something?)
bit of freeze discharging even though soc is 4%
![image](https://github.com/springfall2008/batpred/assets/142018870/337f2ff9-9963-4fec-a55c-75a5c99d555e)
![image](https://github.com/springfall2008/batpred/assets/142018870/c5d56607-f730-4487-ac39-88f375fd7aa1)

best_soc_keep = 0.5
![image](https://github.com/springfall2008/batpred/assets/142018870/76c7fe5b-d399-4649-862f-d6b447b54843)
![image](https://github.com/springfall2008/batpred/assets/142018870/3ff70862-0a4a-4559-a1dd-98c4504fda6e)

best_soc_keep = 0.6
bit more charging but to 4% limit??
I noticed the import colours all go amber/red and then turn green as the plan is recalculated?
![image](https://github.com/springfall2008/batpred/assets/142018870/3d50e6a7-1aa2-4b01-b237-50746cffb569)
![image](https://github.com/springfall2008/batpred/assets/142018870/fd92b555-181e-4c16-9a06-44747c8dc881)

**Comment by gcoan** [comment]
best_soc_keep = 1.0, still the same
![image](https://github.com/springfall2008/batpred/assets/142018870/015e838a-28de-46c4-8923-12a3dd587dec)
![image](https://github.com/springfall2008/batpred/assets/142018870/86657966-1b7b-44ff-8e45-c6f3a921263c)

best_soc_keep = 1.0, but dropped charge & discharge loss from 5% to 4%.  One more charging slot has appeared but again only to 4% limit
![image](https://github.com/springfall2008/batpred/assets/142018870/a842f9ec-0743-4632-ac1a-93cfe9b43930)

**Comment by springfall2008** [comment]
Can you try this one? https://github.com/springfall2008/batpred/releases/tag/v7.13.13

**Comment by gcoan** [comment]
I had set battery losses back to 0.05 and best soc keep to 0.3 and turned predbat back on as I was happy with the two charges planned for this evening 

Here's the latest plan on 7.13.11
![image](https://github.com/springfall2008/batpred/assets/142018870/e356f72a-baef-4aa8-a168-c1986c3b94ec)
![image](https://github.com/springfall2008/batpred/assets/142018870/5c88ae5d-e356-480a-819a-b345b4e25c7b)

Then installed 7.13.12 and positive improvement, charging again overnight 
![image](https://github.com/springfall2008/batpred/assets/142018870/ee66056e-91bc-42b4-9a47-1ecc6c546252)
![image](https://github.com/springfall2008/batpred/assets/142018870/1cc17eaa-57e8-41f6-b580-082b08730bb0)
![image](https://github.com/springfall2008/batpred/assets/142018870/b8ce1c34-4c6f-4d7a-b196-5305edbd0f8e)
![image](https://github.com/springfall2008/batpred/assets/142018870/ec76772e-7650-49fb-9bd9-5b47368614d7)

Unfortunately at the moment no recharge before the evening peak tomorrow, but this might emerge nearer the time

**Comment by gcoan** [comment]
7.13.13

Looks fairly similar to.12, charging in the slightly cheaper slots this evening, and overnight tonight and tomorrow night. 
Unfortunately at present no planned pre peak charge tomorrow afternoon 
![image](https://github.com/springfall2008/batpred/assets/142018870/88fbf6c1-6941-458e-a0eb-abbe1b82a337)
![image](https://github.com/springfall2008/batpred/assets/142018870/ccb83cf6-d680-4371-8796-778a59147979)
![image](https://github.com/springfall2008/batpred/assets/142018870/1b288d00-2406-4c4c-afef-744428accae8)
![image](https://github.com/springfall2008/batpred/assets/142018870/576d231b-4e26-4db2-9046-7683eeed5609)
![image](https://github.com/springfall2008/batpred/assets/142018870/ad577e5a-45f6-41b2-ba21-49cecac10d1b)

**Comment by springfall2008** [comment]
The pre-peak charge will come if needed based on the always include the next N hours in the calculation part

**Comment by slopemad** [comment]
I've put 7.3.13 on and here's my plan.  The surprise, for me, is that it is doing it's "bulk charge" between 22:00 and 01:00 and holding at full charge overnight, and therefore charging in those 22, 23 and 24p slots, rather than maybe holding the charge in those more expensive slots (or maybe even discharging), and instead charging in that cheaper 04:00 to 07:00 slot.  Still seeing threshold fluctuations also.

<img width="573" alt="Screenshot 2023-11-25 at 20 24 39" src="https://github.com/springfall2008/batpred/assets/118896510/281fff6c-43c6-41b4-9b84-278ccb2120ee">
<img width="398" alt="Screenshot 2023-11-25 at 20 29 30" src="https://github.com/springfall2008/batpred/assets/118896510/1cb13aa6-1a29-453b-b4ab-1e353638ca62">

**Comment by springfall2008** [comment]
@slopemad I think this is as expected, you avoid charging/discharging/inverter losses when holding at 100% so it's saving maybe 15% which probably outweighs the small rate differences.

That said I'd made another release today (sorry for all the flux) which maybe worth trying: https://github.com/springfall2008/batpred/releases/tag/v7.13.14

It's really hard for me to test in all cases which is why you end up being beta-testers

**Comment by gcoan** [comment]
@slopemad mine doesn't start the charge to 100% until the rates drop at 4am.  Quite different to yours 

Those fluctuations in threshold are all before the dashed line so hard historical aren't they?

**Comment by gcoan** [comment]
Installed 7.13.14

For me.12 and.13 were better, they didn't charge in slightly higher slots at 2330 and 0030 for example that.14 is now continuing to charge in 

![image](https://github.com/springfall2008/batpred/assets/142018870/98c839e3-65af-450c-870f-5d7fe1eaafc0)
![image](https://github.com/springfall2008/batpred/assets/142018870/b4a027e9-e990-4e03-8836-085d3e5bf6a7)
![image](https://github.com/springfall2008/batpred/assets/142018870/ed8c78ef-88ac-49f0-8075-a277cadbb451)

**Comment by slopemad** [comment]
v 7.3.14 - I'm OK with the idea that it will hold it's charge all night in one big massive charge window as that'll likely be cheaper, but surely it should be picking the cheapest slots inside that window to actually do the increasing of battery.  So if it's basically in charge mode from 21:30 until 07:30, and can do about 13% in each half hour slot (so five slots), then it could pick the 5 cheapest slots in that window to do the actual charge.  (The logic for charging the car is perfect in this respect, it picks the cheapest slots to achieve the desired SOC by a certain time).

There was no fundamental difference in plan with calculate second pass on or off, or combine charge slots on or off.

<img width="578" alt="Screenshot 2023-11-25 at 21 27 10" src="https://github.com/springfall2008/batpred/assets/118896510/3f3899cf-995c-4820-8129-02d3b90e9a70">

**Comment by gcoan** [comment]
> Installed 7.13.14
> 
> For me.12 and.13 were better, they didn't charge in slightly higher slots at 2330 and 0030 for example that.14 is now continuing to charge in 
> 
> ![image](https://github.com/springfall2008/batpred/assets/142018870/98c839e3-65af-450c-870f-5d7fe1eaafc0)

Now I look more closely I see the 2330 slot labelled as "charge" is actually (looking at the soc) not charging but holding the charge.  0030 is charging though when the rate is 24p

I still prefer.13 or 12 as the bulk of the overnight charging was planned 0300-0600 when the rates are 19/20p whereas.14 is charging the bulk at 2230-0100 when the rates are 22p

Best on.14 was 2645p

Just downgraded to.13 and best is 2615p so confirms my thoughts

**Comment by slopemad** [comment]
Also, Low/High thresholds are still fluctuating...

**Comment by slopemad** [comment]
OK, it turns out I'd managed to downgrade from 7.3.13 to 7.3.12 rather than up to 7.3.14.  I've now actually upgraded to 7.3.14 - the plan looks a bit more sensible in terms of when it picks charge slots overnight.  Will monitor to see if it sticks to this...

<img width="576" alt="Screenshot 2023-11-25 at 23 01 08" src="https://github.com/springfall2008/batpred/assets/118896510/8a9b3c18-6f3f-44c2-aaa1-35aa81dba487">

**Comment by Redding43** [comment]
Latest version running in mine. Lots of charge slots with little charge activity
![IMG_3156](https://github.com/springfall2008/batpred/assets/122637671/5783f058-faa7-4735-bebe-f4dc8c278599)
![IMG_3157](https://github.com/springfall2008/batpred/assets/122637671/2954e9ba-e1bc-44d8-bfa2-e8bd08976bce)

**Comment by Redding43** [comment]
![IMG_3158](https://github.com/springfall2008/batpred/assets/122637671/11cf0cf8-1ec7-40b1-9b1a-56b7f3fdf17d)

**Comment by gcoan** [comment]
I left .13 running last night as it was a cheaper plan than .14, but it was a bad night as there wasn't much in the way of cheap slots and it was very cold so my ASHP ran most of the night so loads of electricity used already. At least the house is warm.

Battery only got to 80% or so overnight and was 71% this morning. Function of the poor rates.

This morning's plan on .13 
![image](https://github.com/springfall2008/batpred/assets/142018870/45346551-2657-41ee-8139-bb81651a0641)

And as I took the screenshot it was 8:55 so predbat completed a run cycle and the red import slots changed to green - why is this that slots sometimes different colours but the plan is the same?

![image](https://github.com/springfall2008/batpred/assets/142018870/76c16906-b66b-4e23-90a0-dccba29e7dc2)

I then upgraded to.14 and this plan is slightly better as it's now charging in the 9:00 and 9:30 periods at 24/25p, but the charge is then held (and grid imported) through the slightly more expensive 12:00 28p slot and then soc is released from 14:30 when the rates are 26, 25p to end up with the battery completely flat by very early in the peak period.

![image](https://github.com/springfall2008/batpred/assets/142018870/01220899-e55c-4e45-9e8c-b9ea92f0c86b)

Appreciate that with the rates being so flat and battery losses of 5+5% there's very little benefit of discharging and then recharging through the day today, but I really need the soc to be held for the peak period where the rates are 20p higher than the afternoon rates, and not have the soc released prematurely.
I'll leave this running and see if the plan improves and the soc is retained for more of the peak.

Also, can "hold charging" (or is it freeze charging) be added to the plan? It's hard to read eg 11:00 through to 14:00 all say "charging" with an up arrow to 100%, but the soc column shows it's actually already on 100% so it'll be holding the charge.  On first review these slots appear identical to 9:00 and 9:30 slots that actually both are charging to 100%

**Comment by Redding43** [comment]
This is what I have woken up to today, latest Predbat version
![IMG_3160](https://github.com/springfall2008/batpred/assets/122637671/95254f86-7401-4767-9f81-76685706ea15)
![IMG_3161](https://github.com/springfall2008/batpred/assets/122637671/664a0bba-6512-415a-95ad-7e316f390cc8)

**Comment by gcoan** [comment]
The broad thrust of your plan looks reasonable @Redding43.  Your battery isn't being charged during the day as there isn't any low rate slots, the battery will see you through to Monday AM when there are some cheaper charging slots to refill the battery.

#But some of the detailed slot choices seem odd. Why release the charge at 11:30 when import is 25.8p but then hold the charge (and import) in the 12:00 slot when the price is 28p?
Your half hour slot consumptions are tiny so it's not going to make much £ difference but still ...

**Comment by gcoan** [comment]
> I then upgraded to.14 ... but soc is released from 14:30 when the rates are 26, 25p to end up with the battery completely flat by very early in the peak period.
> 
> ![image](https://github.com/springfall2008/batpred/assets/142018870/01220899-e55c-4e45-9e8c-b9ea92f0c86b)
> 
> Appreciate that with the rates being so flat and battery losses of 5+5% there's very little benefit of discharging and then recharging through the day today, but I really need the soc to be held for the peak period where the rates are 20p higher than the afternoon rates, and not have the soc released prematurely.
> I'll leave this running and see if the plan improves and the soc is retained for more of the peak.
> 

Update: 11:15 and the 100% soc is being retained now until 15:30 so now covers more of the peak import period than the plan at 9:00. So as the day advances it's looking better, though since predbat is planning for 36 hours ahead I'd have thought this would have been covered in the earlier plan?

Anyway, looking better
![image](https://github.com/springfall2008/batpred/assets/142018870/ef3ea018-9698-4dc5-a119-eb7e305905c0)

**Comment by Philhews** [comment]
The 4 hour forced inclusion is still making mine do stuff I don't want it to, it just set a hold charge on a 26p slot when the battery can be charged in the future at 19 (losses make it worth using the battery). Previously without the forced 4 hours and with the ability to select a number of charge windows I could control this and prevent the behaviour but now I get no choice.

**Comment by slopemad** [comment]
Looking at the plan on mine, it's holding the battery at 100% until 4pm now (If I was doing it manually, I'd probably stop at 15:30).  And after the peak, it starts charging the battery back up again at some of the more expensive rates rather than the lower rates later in the night.  As I suggested earlier, perhaps when Predbat has a prolonged 'hold charge' block like this, it could then use logic similar to the car charging logic which just selected the cheapest rates in that big block to do the actual charging.

At the moment though, it's very marginal whether being on Agile is worth it versus Cosy, and there's no sigh of improvement in the rates coming up.  (And the Octopus Tracker rate is, for example, 21.4p today and that's without needing to factor in battery losses).  So I'll probably shift to Cosy tomorrow for a while.

<img width="615" alt="Screenshot 2023-11-26 at 12 13 11" src="https://github.com/springfall2008/batpred/assets/118896510/2e1d5be8-efbf-48d2-a443-7f60703afc1a">

**Comment by gcoan** [comment]
@slopemad agree, you've got quite low load for each slot so it would look to be better to not charge until the cheaper rates overnight.

I had been wondering about whether I should change tariff off agile as well as rates are similar to the price cap and I currently only have a single 5.2 battery that is marginal to see me through the peak on a normal day, and definitely isn't enough now it's turned cold.  

Looking on the Octopus Compare app though, Agile was £2 cheaper than Cosy for yesterday and £20 cheaper for the last week. Flux would have been slightly cheaper than Cosy (but only a tiny bit).
Tracker would have been £3 cheaper than agile for yesterday, about the same for the last week and slightly more expensive for the last month.

I'll keep an eye on it but will stick with Agile

**Comment by gcoan** [comment]
15:25 the plan up to and through the peak period is fine, but now more charging is planned for in comparatively expensive rates after the peak, eg 19:00 at 27.5p
![image](https://github.com/springfall2008/batpred/assets/142018870/65ce029e-bcf3-49b1-b139-edda3da6d31f)

I'll let predbat run through the peak (assuming this stays the same) and put it into read-only afterwards for the evening if it keeps wanting to charge.

I know there's lots of comments but this is an improvement over 7.13.10 yesterday where I had to stop predbat from charging through the peak period and manually change to Eco mode

**Comment by springfall2008** [comment]
I've just made a new release which is largely in place to reflect better on the plan what is going on. The HTML plan now has a cost column and the charge column has some new states (NoCharge and HoldCharge). 
There is also a plan debug enable toggle which will show the rates used for planning accounting for losses in brackets.

https://github.com/springfall2008/batpred/releases/tag/v7.13.15

**Comment by gcoan** [comment]
Thanks @springfall2008 I'll try the .15 version.

Update on .14 through the peak period, predbat kept the battery held at 100% through to 16:30 as planned and then released it for discharging which its now doing.
Interesting the "charging immediately after the peak period at unfavourable rates" I'd reported earlier has advanced backwards, almost as a rolling hour by hour window, so now at 17:40 its now not planning to start charging until 21:00.
![image](https://github.com/springfall2008/batpred/assets/142018870/18747b2b-99fa-46b5-ae7b-f83b63cdfbc7)

**Comment by gcoan** [comment]
Installed .15
![image](https://github.com/springfall2008/batpred/assets/142018870/692ee835-635e-4e8d-ba1b-3415c60c59e6)

Good things:
- addition of the £ value is useful, although I don't like the values its now telling me!
- inclusion of hold charge (does this now replace freeze charge?)
- from 22:00 the plan looks good. Its charging in the cheaper slots and discharging in the more expensive.  Hold charging eg 00:30 where the rates are higher but not enough to make it worthwhile to discharge with battery losses
- pre-charging from 13:30 to 15:30 ahead of the 4pm peak tomorrow (although there still is unnecessary post-peak charging planned)
![image](https://github.com/springfall2008/batpred/assets/142018870/0567ce17-9938-456d-98f8-86ff103d9c7d)

Minor issues:
- cost display over midnight.  Costs are shown I believe as today's costs from now, but don't reset to zero at midnight.  I know this is consistent with the way the rest of predbat works but is confusing as the cost to Tuesday 05:00 (last row of my plan) is all of Sunday's costs+all of Monday's+5 hours of Tuesday
- still something weird going on with colouring of the import rates.  They've changed colour between red, yellow and green in subsequent runs of predbat.  e.g. Mon 07:30-10:00 was Green in the plan at 17:55, then at 18:00 turned Yellow
![image](https://github.com/springfall2008/batpred/assets/142018870/e2be594b-0814-45ad-bc4d-d5c4b49747ca)
and at 18:05 turned back to green!
![image](https://github.com/springfall2008/batpred/assets/142018870/223abc76-12c5-4184-adb6-90a506f58bfe)

- the charging at 21:00 and 21:30 tonight isn't necessary.  May be a hang-over of the rolling 'forward window after peak period' I identified above

I took a look at the debug version of the plan as well.  Feel that you have to be careful how you interpret it.  The "with cost" figures only of course apply if you are charging or discharging the battery.  If the battery is doing nothing (either exhausted or being held) then just the normal import rate applies.  So just need to read it carefully

**Comment by slopemad** [comment]
OK, so 7.3.14 vs 7.3.15

<img width="689" alt="v7.3.14" src="https://github.com/springfall2008/batpred/assets/118896510/a51a84e9-18fc-459e-8789-b287aba1a4c9">
<img width="769" alt="v7.3.15" src="https://github.com/springfall2008/batpred/assets/118896510/f836ed9a-3193-4d16-ab3b-8d7d95890fcf">

So I still can't help but think it could be choosing some slightly cheaper slots to actually fill the battery.  Also, the first few slots are using multiple different states to hold the battery at 64% SOC ;-)

**Comment by DJBenson** [comment]
Agree on the point about the cost column - it would make more sense to most users if that reset at midnight and spanned the single day only, but appreciate why it works the way it does.

**Comment by gcoan** [comment]
The predbat cost just reset to zero as it clocked over midnight which is good.  It's jus5 looking at the plan where it spans days the cost looks odd 

Latest plan is looking pretty good, it's not perfect -the 4:30am slot at 19.4p isn't being used for charging as the battery is already full by then (and this is the cheapest slot of the night) but it's only a couple of pence difference from the slots that are being used 
![image](https://github.com/springfall2008/batpred/assets/142018870/a0aaa173-f5f0-4eb2-950a-cb32d85c7a59)

**Comment by springfall2008** [comment]
What do you think of this HTML where the charge slots are merged when longer and the cost is the actual cost rather than the total?

![image](https://github.com/springfall2008/batpred/assets/48591903/465eb507-3ab5-4ad1-9aab-71402b8aea9a)

**Comment by DJBenson** [comment]
I like that the above includes _something_ for the previously blank cells, gives confidence that it is doing something. Also like the merged cells. Not sure on the cost column - it now shows the cost of that particular slot? Might be good to have slot cost and accumulative cost to satisfy both camps.

**Comment by springfall2008** [comment]
Released here, I split the incremental cost out into a different column: https://github.com/springfall2008/batpred/releases/tag/v7.13.16

**Comment by slopemad** [comment]
7.3.15 (top) vs 7.3.16 (below)...

Interesting, it didn't combine the 21.30, 22.00 and the 22.30-07.30 charge sessions, or the 14.30-15.30 charge sessions.
As mentioned before, it would be useful if when it does have adjacent charge sessions like this, that it would pick out the cheapest slots in that to actually top up the battery, rather than topping it up early on in the session and holding it there.

Also, as you can see between 15.30 and 16.30, it just has a 'FreezeChrg' session, but given that the second half hour is 8-9p more expensive than the first half hour, it would surely be better for it to top up the battery between 15.30-16.00 and just running off battery from 16.00 onwards?

<img width="515" alt="v7.3.15" src="https://github.com/springfall2008/batpred/assets/118896510/2031a0a9-a788-416f-975e-954443d9c5ee">
<img width="531" alt="v7.3.16" src="https://github.com/springfall2008/batpred/assets/118896510/cfe958a3-ac38-438a-b171-dee3887d65cb">

**Comment by springfall2008** [comment]
I don't quite get that 32p slot at 16:00, I'd tend to agree it seems odd and I'm not even sure the freeze was required. I wonder if it goes away or sticks around. If it's staying can you capture the log?

**Comment by Philhews** [comment]
I like the merging of slots, makes things clearer, also like the cost per slot, slight issue is that on my phone the html is too wide to fit and I can't scroll it, anyone know how to resolve that? If the answer is something with card mod then I'm happy to go that route(font size change or something) 
![Screenshot_2023-11-27-14-07-31-31_c3a231c25ed346e59462e84656a70e50](https://github.com/springfall2008/batpred/assets/117396644/7529986e-3c1e-4f0f-bc52-d0270a51ed3c)

**Comment by gcoan** [comment]
I have noticed that there's a tendency for predbat to plan things that "don't look right" but when it gets to them, the plan changes for the better

Eg this morning 6am, the 7:30 and 8:30 slots were both planned to hold the soc at 100% despite these being more expensive and there were cheaper slots later on where the battery could be recharged
![image](https://github.com/springfall2008/batpred/assets/142018870/83693434-a0ce-4606-ab1b-e71ee060e340)

But by 8:30 the plan had changed and the battery was discharging in this higher price slot
![image](https://github.com/springfall2008/batpred/assets/142018870/4283d203-9a39-4eca-ac75-c8bbef495f2a)

I've just upgraded to.15 and the plan looks pretty good apart from it should b3 discharging at 18:30 not holding the charge.
But this may change when we actually get there
![image](https://github.com/springfall2008/batpred/assets/142018870/24d68252-8fca-4712-824c-626ae7bc3c93)

I like the combined blocks and the explicit down arrow for battery discharging. I'd still like to see the cost today resetting to zero at midnight in the forward plan.

@Philhews I find that if I rotate my phone to horizontal I can see the multiple columns of the predbat plan better

**Comment by Philhews** [comment]
The problem then is my phone thinks the dashboard has space for 2 columns and so the plan doesn't get any more space. Time to play with a vertical stack card maybe, last time that didn't display the plan card at all though.

**Comment by gcoan** [comment]
I’m using the grid layout of the layout-card HACS front end integration.  I found this displays the charts best on my iPad & phone

**Comment by JonathanLew1s** [comment]
I've set Low Rate Threshold to 0.8 and see a more acceptable low rate band. 

![image](https://github.com/springfall2008/batpred/assets/34936213/63e1c878-0893-4e66-b098-2067be6cb65b)

**Comment by gcoan** [comment]
Couple more examples of the plan planned changing to something more sensible as it gets nearer to the execution time

At 14:40 shouldn't hold the charge at 18:30 or charge at 19:00
![image](https://github.com/springfall2008/batpred/assets/142018870/b3e89026-25d7-443e-9f50-8e0bfbc3a231)

At 15:45 the 18:30 charge hold now isn't there and the battery is discharging, but the 19:00 charge still present 
![image](https://github.com/springfall2008/batpred/assets/142018870/9158ec88-5838-42dd-8ea1-0f997623fb1a)

At 16:55, 18:30 is discharging and the charge at 19:00 has gone as well, giving a sensible plan
![image](https://github.com/springfall2008/batpred/assets/142018870/2ccb91a9-f46b-463f-b3ef-baabe2a4aa95)

**Comment by gcoan** [comment]
@JonathanLew1s I tried my own experiments with changing rate low threshold and (at the moment) 0.8 does seem to be the sweet spot.
My only concern is whether (as per above) predbat changes the plan when it gets to execute it and we end up with a worse plan as a result?  Certainly in terms of predicted forward activity it looks much more sensible than the default 0 plan.

**Rate low threshold on the default of 0**, **best is 2341**
![image](https://github.com/springfall2008/batpred/assets/142018870/f3a45e86-f79e-4a41-84c7-b2f03aab8433)

Changing rate low threshold to **0.9**, best is slightly reduced to **2308**
![image](https://github.com/springfall2008/batpred/assets/142018870/f20e8c44-3e84-49b0-b15b-047402377fce)

Reduced rate low threshold to **0.8**, best is reduced further to **2252**
![image](https://github.com/springfall2008/batpred/assets/142018870/9197cddb-cb85-4c49-92d4-8c02c0080e13)

Rate low threshold changed to **0.7**, only one charging slot overnight now, and best increases to **2329**
![image](https://github.com/springfall2008/batpred/assets/142018870/0c3ade31-f531-4493-ac2c-96df8f8bfe2f)

**Comment by Philhews** [comment]
With experiments around the low rate threshold (rather than leaving on auto) I wonder if the max charge windows parameter needs to come back in auto mode, that way you can control the number of slots rather than guessing what the rate spread is vs the average to pick a low threshold.

**Comment by Redding43** [comment]
Minor comment but the accumulative cost does not work with dark mode. Text colour means it can’t be read
![IMG_3164](https://github.com/springfall2008/batpred/assets/122637671/5956571a-bcd3-4e70-a42d-9b8571c2b2a7)

**Comment by gcoan** [comment]
Hmm, just realised that rate low threshold set to 0.8 is NOT charging the battery before tomorrow's peak (at ouch, 53pkW rates)
![image](https://github.com/springfall2008/batpred/assets/142018870/a5b47699-b91d-44ac-9438-4e5f63293546)

And neither does setting rate low threshold to 0 either
![image](https://github.com/springfall2008/batpred/assets/142018870/26b922f1-86d2-4eb9-bddd-1f51018eeb49)

So gonna have to hope that this is another one where predbat works out what the right thing really is nearer the time ...

**Comment by gcoan** [comment]
@JonathanLew1s 
rate low threshold=0.8 is starting to cause some strange updates to the plan.  At 5pm it had identified to start charging tonight at 22:30pm which is the cheapest slot of the night

But now at 21:00, those early cheap slots are now "hold charge" and the charging is being done from 23:30 when the rates are higher!
![image](https://github.com/springfall2008/batpred/assets/142018870/7758cc8e-355c-4231-95a2-f1b5a1edb709)

Putting it back to rate low threshold=0 because although the long term plan doesn't always look right, predbat has (so far) sorted itself out and charged at the right times when it is near to the slot
![image](https://github.com/springfall2008/batpred/assets/142018870/3a23c3d3-0507-49b7-8e33-c0b69ecb0a91)

**Comment by slopemad** [comment]
Predbat has, for me, come up with a decent overnight charging plan (I think I'd only change one thing, and have it charging the battery at 05.30 in one of the cheap rates, but I think that would save me approximately 1p in total, but also what I find is that when it's 'holding' a charge, the inverter is actually regularly cycling through charging and discharging small amounts, so actually it's probably better in this scenario to get to 100% and keep it there.  I'm not sure Predbat is thinking like that though ;-)

As for the daytime tomorrow, virtually no charging at all which is OK because I'm expecting wall to wall sunshine and about 10kWh of solar.  Only comment in this is that the battery will be fully charged overnight at 18-20p, and then exporting solar at 15p, which I don't think it necessarily optimal, but again it'll probably be OK.  Solar power exported straight out at 15p rather than put into the battery and then discharged from the battery with the losses involved in that, probably irons it all out.

Yes, I still haven't pulled the trigger to switch to Cosy ;-)

<img width="966" alt="Screenshot 2023-11-27 at 23 49 12" src="https://github.com/springfall2008/batpred/assets/118896510/20c704b5-1520-4c0f-9680-4529e6b63806">

<img width="739" alt="Screenshot 2023-11-27 at 23 53 04" src="https://github.com/springfall2008/batpred/assets/118896510/c4507948-057e-472f-82b6-7b6faaece527">

**Comment by springfall2008** [comment]
Can someone else please test out changing 'plan_forecast_hours' to 24 in the HA GUI (expert mode required)? I think it maybe more sensible to only count the next 24 hours.

**Comment by gcoan** [comment]
I left 7.13.16 running last night (with rate low threshold =0) and battery was fully charged overnight.
Similar pattern today of the forward plan not looking optimal in terms of charging, but as the day progresses and i5 gets nearer to the charge slots, the plan corrects itself

9am the plan was to hold the battery until 2pm, then start releasing it, resulting in the battery being exhausted in the middle of the peak period
![image](https://github.com/springfall2008/batpred/assets/142018870/352ead08-5699-4259-9b02-b10fbc4417cc)

12:45, the hold charging retains the battery now to 17:00, meaning importing for the most expensive peak rate slot at 16:30 and the battery still having charge at the end of the peak period
![image](https://github.com/springfall2008/batpred/assets/142018870/cc0c4c39-4b4f-4ee9-9e6e-0774f57eaebb)

And now at 14:25, predbat has now correctly identified 16:30 as being where the battery needs to be releasing charge to avoid peak peak import.  I think 16:00 ought to be the same, releasing charge, trading this for retaining charge to support the 18:30 slot, but at least it's only 4p difference in rate
![image](https://github.com/springfall2008/batpred/assets/142018870/23be7f7f-2033-4cd4-8919-b75216a653f2)

@JonathanLew1s I'll compare settings. I know my metric future rate offset is zero though.
@springfall2008 Ill try changing plan forecast hours from 36 to 24.  I was thinking of doing this anyway as my battery isn't enough to support a whole day anyway.

Tomorrow's peak forecast prices 56, 58 and 63p. Argh !

**Comment by gcoan** [comment]
14:37, plan was the same as 14:25, retaining the charge through 16:00 slot, only starting discharging at 16:30
![image](https://github.com/springfall2008/batpred/assets/142018870/18e9295f-5f9d-4106-b914-aad7e6f7c308)

By 14:49 though the plan had changed so 16:00 is now the start of discharging which is what I'd expect it to be.
Not sure what "no charge" is as there's no PV at 17:00 and the soc is shown as dropping still
![image](https://github.com/springfall2008/batpred/assets/142018870/f59e2877-d508-45de-883e-40b348b9c6c0)

Changed Plan forecast hours from 36 to 24 and TBH looks much the same
![image](https://github.com/springfall2008/batpred/assets/142018870/4444fe9f-e1c3-48d9-858d-80df10221525)

19:00-20:30 is marked as charge but soc limit of 14 means it's holding flat whereas with a 36 hour plan it was shown as idle but with a 14% limit so effectively the same

Best soc keep is 0.3 so don't know why it's retained at 14%

**Comment by JonathanLew1s** [comment]
@gcoan I do not understand why you get 46p as a low rate cost. My low rate threshold is calculated at 20.2p

48 Hour. Future costs based on nordpool with a metric future rate offset in addition. Charge is held on low rate slots where charging is not required, avoiding conversion losses over next day and half. Only have to start questioning past 10pm tomorrow when choice of charge slots is dubious, expect that to firm up before those slots.

![image](https://github.com/springfall2008/batpred/assets/34936213/9ffb2685-9fa2-4a0d-8160-ca5ca6363245)

24 hour. Does have a positive impact in cost, the system is less aggressive with it's charge tonight, maintaining a lower battery target. 

24 hours - risk if tomorrows rates are higher
48 hours - risk if tomorrows rates are lower
Trying 36 :D

![image](https://github.com/springfall2008/batpred/assets/34936213/f501a83c-30ad-4d88-8e49-edc706e59161)

**Comment by slopemad** [comment]
Don't worry about the peak prices, there's a savings session paying £4/kWh saved ;-)

**Comment by JonathanLew1s** [comment]
@gcoan it's strange how we have such varying outcomes from similar settings, only significant points are your usage is much higher due to ASHP and I have more storage. The way I understand your settings, your PV is more optimistic, your future import rates are equal to todays' rates, you plan for a shorter window. If battery sizes were equal I'd expect your settings to charge less.

As an aside, do you know you can automatically join the Octoplus sessions via the Bottlecapdave Octopus integration? Sends a nice notification in HA also.

**Comment by gcoan** [comment]
It's still not forward planning right, plans to charge the battery up ok overnight but then let's the charge release during the day leaving me empty for the peak period (and tomorrow's saving session)

Here's the plan at 22:45
![image](https://github.com/springfall2008/batpred/assets/142018870/74202ded-a7fd-4575-a195-35aa59410401)
![image](https://github.com/springfall2008/batpred/assets/142018870/29bca581-e232-4aa8-9486-d931ac139821)

I then uninstalled predbat, restarted HA and all the addons, manually made all the required changes to apps.yaml and the new plan looks much the same
![image](https://github.com/springfall2008/batpred/assets/142018870/e609708d-bdf8-45fb-a0a4-ed67bf865b3c)
![image](https://github.com/springfall2008/batpred/assets/142018870/0db24369-2774-4b3b-ab69-15b34f885d3b)

I had an issue that despite predbat recognises saving sessions it wasn't planning for them in the import and export rates. Deleting and setting predbat up again has fixed the saving session issue but not the wider forward planning problems. 
I did notice that despite uninstalling and restarting everything, all of the predbat control entities such as best soc keep were retained and I didn't have to populate those again??

**Comment by slopemad** [comment]
I think it's kind of expected, I'm seeing similar now.  Here's my plan from 06:30 until 18:30 - it isn't planned to pre-charge the battery, and therefore there wont be so much to discharge, and the battery SOC will actually be reasonably high prior to the 'red' prices.  But i'm fairly confident that it'll sort it itself out when the time comes.  (That said, I'll keep an eye on it and charge manually between 1500 and 1530 if I think it needs it.  Better to boost at 34.4p and try and avoid all those nasty red prices.)

<img width="662" alt="Screenshot 2023-11-28 at 23 35 34" src="https://github.com/springfall2008/batpred/assets/118896510/f5bd2f49-af27-406d-8ba3-faa6dbf25906">

**Comment by gcoan** [comment]
My predbat plan just isn't taking account of the higher rates and I've had to set it to read only and initiate a force charge whilst the rates are still comparatively low

1am plan last night, charge up overnight, discharge a bit in the breakfast price slots, then charge back up again in the morning - all good
But then let the battery discharge before the evening peak and the saving session- not good but like @slopemad I was prepared to see it develop 
![image](https://github.com/springfall2008/batpred/assets/142018870/a02310df-d17a-44d1-bb0a-219d52d215e7)
![image](https://github.com/springfall2008/batpred/assets/142018870/49d982b9-1f4c-407f-9629-589c7c1de617)

8:40; plan much the same, battery charged overnight, coming up to recharging from 9am - good
Afternoon discharge and nothing for the peak - still not good
![image](https://github.com/springfall2008/batpred/assets/142018870/26e1be6c-c826-42e5-a6d3-14a18a9ae018)

9:02; predbat has decided against charging the battery at 29.5 and 28.6 which are the cheapest slots, instead letting the battery discharge to empty and recharge a bit at 31/32 before leaving empty for the peak - all bad
![image](https://github.com/springfall2008/batpred/assets/142018870/b3107a3c-9a6e-44ef-b1d8-23598b6211d8)

9:17; installed the newest version released this morning
Double checked calculate best discharge is enabled and set discharge freeze only is disabled as per the updated docs - they're both correctly set
Predbat now deciding to hold the charge at 23% but soc is 49% so it's actually still discharging in this cheaper slot - all bad
![image](https://github.com/springfall2008/batpred/assets/142018870/02cc45f2-9250-46b9-ae9f-53d55cbcd76f)

Since I would prefer to be recharging the battery at these 29/30 rates rather than the 33's just before the peak, I set predbat into read only mode and started a charge. Predbat still sticking to its same plan
![image](https://github.com/springfall2008/batpred/assets/142018870/70925e8b-5f9d-46ae-9e8c-c07b320acea4)

**Comment by slopemad** [comment]
Obviously you are seeing different results to me, mine wasn't forecasting to charge at the 'lower' rates this morning, because it was expecting the solar to take care of the load and keep the battery full anyway.  As it happens, because it's so cold, the heat pump has used quite a lot more power than it has been doing and with in day adjustment, it now expects a higher house load so it has now picked up on more charging options and sorted out a reasonable plan to make sure there is enough to discharge in all three slots.  The plan it has come up with for me is actually really pleasing.

I use a pretty aggressive In Day Adjustment damping factor of 1.3 which i find does a good job of working out when my heat pump is going to pull more power (after all, if it's pulling more power overnight, then it'll probably pull more power all day).  In-Day adjustment chart below for reference.

Also, if you think today's rates are bad, just wait until you see tomorrow.  Expecting the low price between breakfast and teatime to not go below about 36p.  It's a good job the sun is shining.  And there might not be a DFS tomorrow to compensate.

<img width="665" alt="Screenshot 2023-11-29 at 10 36 50" src="https://github.com/springfall2008/batpred/assets/118896510/2008d097-eb4c-4891-9078-72ffb5f45391">

<img width="387" alt="Screenshot 2023-11-29 at 10 41 40" src="https://github.com/springfall2008/batpred/assets/118896510/80ca516a-5488-451f-9eef-6d3382678ad4">

**Comment by springfall2008** [comment]
@gcoan can you share a log?

**Comment by gcoan** [comment]
@springfall2008 Here's the logfile from around 9am 
[appdaemon.log.1 (2).txt](https://github.com/springfall2008/batpred/files/13500465/appdaemon.log.1.2.txt)

Latest plan is now looking better after I manually took over and charged the battery up. 
![image](https://github.com/springfall2008/batpred/assets/142018870/f0ba43c5-007c-4abc-acca-f321eeac7c72)

Latest logfile:
[appdaemon (19).log](https://github.com/springfall2008/batpred/files/13500455/appdaemon.19.log)

I've just had a mare of the last 2 hours.  GivTCP _was_ working fine, and then at about 9:30 just stopped reading from inverter 2 which is the one with the battery connected.  Repeated errors reading from the inverter. Inverter was online in the portal, givenergy app, but GivTCP wasn't populating MQTT.
Tried everything, restarted mosquitto, givtcp, the inverter, HA and all the add-on's, physically turned the inverter off and on again, but still same issue. 
Restoring from the GivTCP 2.4.1 backup fixed the issue but why would it work fine for a week and then suddenly not work any more?

**Comment by slopemad** [comment]
@springfall2008 There's a bug in a calculation somewhere, have a look at 16:30

<img width="666" alt="Screenshot 2023-11-29 at 13 06 38" src="https://github.com/springfall2008/batpred/assets/118896510/f5ff9f1f-b748-4063-8f93-15063b712706">

The cost column says -28p which is going to be incorrect for the plan - there's no discharging scheduled for that slot (and I doubt I'll be exporting a couple of kWh in half an hour from a 2.6kWh inverter, when it's dark).

**Comment by slopemad** [comment]
Looking good on 7.13.19.

I assume that it's current plan to charge throughout the evening will go away closer to the time, when it starts disregarding the best soc keep, and just run off the grid with a flat battery.

<img width="672" alt="Screenshot 2023-11-29 at 14 39 27" src="https://github.com/springfall2008/batpred/assets/118896510/00c91282-7c71-4adf-8824-cc531710d1eb">

**Comment by gcoan** [comment]
Plan on 7.13.18.
Spurious charge at 18:30 at a high rate, only to discharge in the following (cheaper) slot
![image](https://github.com/springfall2008/batpred/assets/142018870/2d43fc9e-72be-44b7-bc58-5a6ca982c8aa)

7.13.19 is better, it's got rid of that charge at 18:30

16:30 slot is odd, it says "hold charging" but the soc limit is 81% and the soc is dropping over the period so it looks more like it's actually discharging?
I'd have thought it would be better to hold 100% charge at 16:30, even though it's a high import rate, and keep the 1.54kW of predicted load so that it can be exported in the saving session period?
![image](https://github.com/springfall2008/batpred/assets/142018870/be361223-eb73-42ba-9e9a-670b0c723662)

**Comment by gcoan** [comment]
15 minutes later and predbat is now back to predicting that it'll charge at 18:30 and 19:00 when the rates drop immediately afterwards !

Here's the logfile
[appdaemon (20).log](https://github.com/springfall2008/batpred/files/13502204/appdaemon.20.log)

For now I'm leaving it in read only mode

**Comment by springfall2008** [comment]
> 15 minutes later and predbat is now back to predicting that it'll charge at 18:30 and 19:00 when the rates drop immediately afterwards ! ![image](https://private-user-images.githubusercontent.com/142018870/286650431-53aabd65-57f0-4233-b157-0743ff70610a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNzgyNDAsIm5iZiI6MTcwMTI3Nzk0MCwicGF0aCI6Ii8xNDIwMTg4NzAvMjg2NjUwNDMxLTUzYWFiZDY1LTU3ZjAtNDIzMy1iMTU3LTA3NDNmZjcwNjEwYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTEyOVQxNzEyMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xM2VjM2UwZmMwNzY4ZWQzMTUyYmMwOTUzZjIyYWZmMzNiYzQ3ZmRjNjllZGQwYTBjZDVlYzA4YzAyZDRkMWE0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.lKli1mMHINEJR7NonzjZw62Erv6eBQx5yQn1Ws6y_50)
> 
> Here's the logfile [appdaemon (20).log](https://github.com/springfall2008/batpred/files/13502204/appdaemon.20.log)
> 
> For now I'm leaving it in read only mode

Why no force export?

**Comment by gcoan** [comment]
To clarify: I left predbat in read only mode as it wasn’t devising an optimal plan. It wanted to start discharging to house load before the saving session, thus leaving less available to support me through the saving session or for a forced export.

Just before the saving session started I turned the ASHP off and enabled predbat again.  We’re sitting in darkness with candles (Mrs’s idea not mine!) and predbat is behaving fine now, we’re using very little for house load which the battery is supporting and as we are under the predicted load predbat has started planning for a forced discharge of the excess battery charge.

My issues above were all about the predbat planning this morning and in the run up to the peak period.

**Comment by gcoan** [comment]
The saving session went reasonably well, predbat changed us to Eco mode at the start and then a force discharge part way through.

But after the saving session had ended at 19:15, predbat started charging the battery up despite it being a 32p slot, there being soc in the battery (no min soc) and the following slots all being in decreasing rates.
I had to turn predbat back into read only mode and change the unnecessary charge to being a discharge.

It really seems that predbat struggles to plan with my 5.2 battery and high load of >1kW/half hour slot.
Even within the 4 hour period of min soc not being applied, predbat is doing strange things and I have to intervene.

I think I'll stop posting screenshots and log files and give @springfall2008 a chance to look at what I've posted already

On the positive front we've got a 2 hour octopus power up event tomorrow; free electricity from 11-1pm.  And then the rates go up again

**Comment by springfall2008** [comment]
@gcoan I think you need to set your soc keep to 0 as your battery is just too small to keep a charge level and be optimal in terms of costs. Have you tried that and does it work?

**Comment by gcoan** [comment]
@springfall2008 I hadn't tried that.  Previously when I had more battery capacity I had some strange results so you advised to set soc keep to >4% to resolve, but certainly happy to try as I was pretty happy with predbat before, its just got really difficult to use with the more recent versions, much higher consumption with the ASHP in cold weather and my losing the 9.5 battery.

Here's the plan with best soc keep = 0.3
![image](https://github.com/springfall2008/batpred/assets/142018870/4afbad24-8994-4055-adf3-7301ed0a0e47)

And changing best soc keep to 0:
![image](https://github.com/springfall2008/batpred/assets/142018870/cdbb1734-a1c2-4600-bc68-0aa9a27c9e8f)
![image](https://github.com/springfall2008/batpred/assets/142018870/8bf15bb1-1302-47d1-8779-af94216ab7c9)

This is looking better.  The charges at 19:00, 19:30 and 20:00 when the rates are dropping (and it would be better to just grid import) have gone.   There's a charge at 22:30 when the rate is 5p lower than the following slot, which is fine.
Maybe not charge at 23:30 when the rate drops further overnight, and hold the charge at 15:30 on Friday in return for releasing the charge at 17:30 at a higher rate

But overall this looks definitely better, thanks.   I'll run with this for a while and see what happens

**Comment by gcoan** [comment]
Thanks for the suggestion about dropping best soc keep to zero @springfall2008 that has simplified the plan and eliminated "rogue" charging. It makes sense now I think about it as to why predbat kept on wanting to charge even in poor charge rates, it was to maintain the soc. It seemed to be more aggressive in charging than just keeping 0.3kW in the battery, but anyway the plan execution is better.  The long range plan often looks a little strange but it corrects as it gets to the execution time.

Through the saving session we really dropped the house demand and I kept an eye on predbat as to what it did.

Generally it did the right thing, but not always, so sharing to see if further code improvements are needed or "it's just my expectations".

1. Predbat charged the battery up to 100% before the peak period and saving session which is great but then the plan it produced appeared to start releasing the soc from the 16:00 slot when import rate was 57p.
Given that the historical load through the saving session is higher than the size of my battery, my battery would be exhausted during the saving session, thus incurring the 457p import rate. This didn't seem right.
![image](https://github.com/springfall2008/batpred/assets/142018870/8e79edfb-8d0c-4a61-ae2c-25542b9fc68b)

2. At 15:10 predbat changed to Hold charging (reserve 100% still) and a little while later I noticed that the battery was discharging. I did a force charge to recover the lost charge before the saving session, and a bit of testing I discovered that even if reserve is 100%, if there is a discharge period covering the current time period, the battery will discharge:
![image](https://github.com/springfall2008/batpred/assets/142018870/82bef3c7-d708-48ff-ba9b-ec5d092ae1f7)

I suggest predbat needs to check and if necessary change the discharge time slots if it's setting charge hold mode.

3. After the saving session and peak period, predbat seems to have given up planning any overnight recharging of the battery. I thought this might have been a hangover from the peak rates of the saving session, but at 19:35 it's still only planning for two half hour charging slots, nothing overnight despite th3 rates falling to 18p and mos5 of Saturday is 28p 🤷‍♂️
![image](https://github.com/springfall2008/batpred/assets/142018870/e61ff00f-f18f-45f7-a0b1-bcbdc7dd1722)

**Comment by gcoan** [comment]
Further to above comment, here's the logfile for 15:00-16:00

[appdaemon.2.log](https://github.com/springfall2008/batpred/files/13532200/appdaemon.2.log)

And the most recent logfile.  Still no meaningful overnight charging planned.  Looked at all the settings. Other than set charge freeze being disabled, and I re-enabled it (no change), couldn't see anything wrong.

[appdaemon (21).log](https://github.com/springfall2008/batpred/files/13532209/appdaemon.21.log)

I did notice from the charts that high and low theshold are both set to 16.1 which I'm guessing is the cause of it not charging, but metric min improvement is 0, metric min improvement discharge is 0.1, rate high threshold and rate low threshold both 0
![image](https://github.com/springfall2008/batpred/assets/142018870/3e0be127-2990-4f37-a54e-71df8cbf7650)

**Comment by springfall2008** [comment]
I think this is resolved now by the newer first pass optimisation algorithm. Please open a new ticket for any new issues.
